src/Symfony/Component/String/AbstractString.php

```diff
@@ -0,0 +1,737 @@
+    {
+        if (!$this->endsWith($suffix)) {
+            return $this->append($suffix);
+        }
+
+        $regex = '{(?:'.preg_quote($suffix).'){2,}$}';
+
+        return $this->replaceMatches($regex.($this->ignoreCase ? 'i' : ''), $suffix);
+    }
+
+    /**
+     * @return static
+     */
+    public function ensureStart(string $prefix): self
+    {
+        if (!$this->startsWith($prefix)) {
+            return $this->prepend($prefix);
+        }
+
+        $next = clone $this;
+        $next->ignoreCase = $this->ignoreCase;
+
+        do {
+            $str = clone $next;
+        } while ('' !== $next->string = $next->after($prefix)->string);
+
+        return $str;
+    }
+
+    /**
+     * @param string|array $string
+     */
+    public function equalsTo($string): bool
+    {
+        foreach ((array) $string as $string) {
+            if ($this->equalsTo((string) $string)) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    /**
+     * @return static
+     */
+    abstract public function folded(): self;
+
+    /**
+     * @return static
+     */
+    public function ignoreCase(): self
+    {
+        $str = clone $this;
+        $str->ignoreCase = true;
+
+        return $str;
+    }
+
+    /**
+     * @param string|array $needle
+     */
+    public function indexOf($needle, int $offset = 0): ?int
+    {
+        $i = \PHP_INT_MAX;
+
+        foreach ((array) $needle as $needle) {
+            $j = $this->indexOf((string) $needle, $offset);
+
+            if (null !== $j && $j < $i) {
+                $i = $j;
+            }
+        }
+
+        return \PHP_INT_MAX === $i ? null : $i;
+    }
+
+    /**
+     * @param string|array $needle
+     */
+    public function indexOfLast($needle, int $offset = 0): ?int
+    {
+        $i = null;
+
+        foreach ((array) $needle as $needle) {
+            $j = $this->indexOfLast((string) $needle, $offset);
+
+            if (null !== $j && $j > $i) {
+                $i = $offset = $j;
+            }
+        }
+
+        return $i;
+    }
+
+    public function isEmpty(): bool
+    {
+        return '' === $this->string;
+    }
+
+    /**
+     * @return static
+     */
+    public function join(array $strings): self
+    {
+        $str = clone $this;
+        $str->string = implode($this->string, $strings);
+
+        if (!preg_match('//u', $str->string)) {
+            throw new InvalidArgumentException('Invalid UTF-8 string.');
+        }
+
+        return $str;
+    }
+
+    public function jsonSerialize(): string
+    {
+        return $this->string;
+    }
+
+    abstract public function length(): int;
+
+    /**
+     * @return static
+     */
+    public function lower(): self
+    {
+        $str = clone $this;
+        $str->string = mb_strtolower(str_replace('İ', 'i̇', $str->string), 'UTF-8');
+
+        return $str;
+    }
+
+    public function match(string $pattern, int $flags = 0, int $offset = 0): ?array
```

Comment:
This return type could use a doc comment to show what is inside of the returned array.

Reply:
added